kernel/src/transaction/mod.rs

```
@@ -295,32 +315,58 @@ impl Transaction {
             ));
         }
 
-        // validate domain metadata
-        let mut domains = HashSet::new();
-        for domain_metadata in &self.domain_metadatas {
-            if domain_metadata.is_internal() {
+        // validate user domain metadata and check if we have removals
+        let mut seen_domains = HashSet::new();
+        let mut has_removals = false;
+        for dm in &self.domain_metadatas {
+            if dm.is_internal() {
                 return Err(Error::Generic(
                     "Cannot modify domains that start with 'delta.' as those are system controlled"
                         .to_string(),
                 ));
             }
-            if !domains.insert(domain_metadata.domain()) {
+
+            if !seen_domains.insert(dm.domain()) {
                 return Err(Error::Generic(format!(
                     "Metadata for domain {} already specified in this transaction",
-                    domain_metadata.domain()
+                    dm.domain()
                 )));
             }
```

These feel like checks that should be done in `with_domain_metadata`:
* `if dm.is_internal()`
*  `if !seen_domains.insert(dm.domain())`

Generally I prefer to fail early. Feel free to make this a followup issue.

---

I generally agree with the fail fast principle, but this case warrants an exception imo.

We're treating these methods as builder methods (return `Self`, even though there is no explicit `build` call) so we should present a fluent builder API. If we do the checking of the error conditions at this point, then we need to return `DeltaResult<Self>`, which is an awkward interface for both Rust (chaining requires try operator), but particularly for engines across the FFI boundary.

wdyt?

---

> which is an awkward interface for both Rust (chaining requires try operator)

I was willing to deal with an awkward interface

> but particularly for engines across the FFI boundary. 

But this is a really good point! GJ